### ISSUE TITLE
[1.12-rc2] docker doesn't handle manually stopped containers well

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#23704]
**test case 1 (no ports published)**

```
docker service create --name foobar --replicas 5  nginx:alpine

# (wait tasks to be up)

docker stop $(docker ps --filter label=com.docker.swarm.service.name=foobar -q)

# (wait for swarm to recover)

docker start $(docker ps --filter label=com.docker.swarm.service.name=foobar -aq)
```

After this, there's 10 containers running, docker doesn't kill/remove the excessive containers

I'd expect docker to kill/remove those containers

**test case 2**

```
docker service create --name foobar --replicas 5 -p 8008:80  nginx:alpine

# (wait tasks to be up)

docker stop $(docker ps --filter label=com.docker.swarm.service.name=foobar -q)

# (wait for swarm to recover - note in my case it didn't fully recover, but hung at 3/5)

docker start $(docker ps --filter label=com.docker.swarm.service.name=foobar -aq)
```

results in:

```
Error response from daemon: Address already in use
Error response from daemon: Address already in use
Error response from daemon: Address already in use
Error response from daemon: Address already in use
Error response from daemon: Address already in use
Error response from daemon: Address already in use
0467a6877366
2ab26cbf3e20
Error response from daemon: Address already in use
fd07976c407f
Error response from daemon: Address already in use
Error response from daemon: Address already in use
79d552f268aa
Error response from daemon: Address already in use
0d659b1bafa1
Error: failed to start containers: 7f4936520a2a, 26a3b00257e1, 616064cf7938, 3c3fcb883565, 9e15f45321e6, b2fcb6182fe7, 8583b7c2dabf, f887d00e4811, 72e0596b161a, 249800a970a0
```

Tested on Docker for Mac, but similar results on other platforms

```
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        Fri Jun 17 20:35:33 2016
 OS/Arch:      darwin/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   a7119de
 Built:        Fri Jun 17 22:09:20 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Comment by ardnaxelarak** [comment]
#dibs

**Comment by tonistiigi** [comment]
> After this, there's 10 containers running, docker doesn't kill/remove the excessive containers

I'm not sure what you mean by "doesn't kill" as you stop the containers yourself. The command you run includes `-a` so I guess you only mean remove.

It may not be a good idea to remove these containers, at least as long as they continue to appear in `docker service tasks -a`. They may be crucial to debugging the issues that caused them to crash. They are cleaned up though when you do `service rm`. cc @stevvooe 

The other part of the report `Address already in use` seems to be a networking issue. This error appears on container create, I assume because the port is already used. Maybe makes sense to track this in a different issue. cc @mrjana

**Comment by mgoelzer** [comment]
Could network partition cause the same problem?

**Comment by mgoelzer** [comment]
ping @aaronlehmann

**Comment by aaronlehmann** [comment]
@mgoelzer: I'm not sure I understand the question correctly. A network partition could lead to a situation where more containers than the desired number are running across all nodes, but this is by design because the alternative is to have situations where no instances of the container are running because the two sides of the network partition assume that the containers are being run on the other side.

The issue here seems to be more about how the agent logic should handle manually stopping and starting containers associated with a service.

**Comment by mrjana** [comment]
@thaJeztah @tonistiigi `Address already in use` is expected because when you stopped the containers which were exposing ports swarm recreated those containers with probably the same IPs so when you restarted the same container it is going to complain about IP address conflicts. It is actually good that it is failing to start the container because it detected an IP address conflict rather than starting them but using an IP of another container. That will be worse.

At the end of the day I think we should have a generic way to deal with the users trying to manage the containers OOB. May be we should not allow stop/starts to happen for managed containers. If the container crashes they can be kept around for debug purposes. But IMO we should block users from stop/starting managed containers

**Comment by tonistiigi** [comment]
@mrjana I thought I got this error already on stop when it was doing the replacing but couldn't reproduce anymore. Maybe it was fixed recently. 

> May be we should not allow stop/starts to happen for managed containers

This seems correct thing to do.

**Comment by thaJeztah** [comment]
A start would be to allow filtering on "managed" / "unmanaged" containers

**Comment by tuesdaythefifth** [comment]
> May be we should not allow stop/starts to happen for managed containers


I don't think blanket changing fundamental docker behaviour (start/stop) is the answer.

Users should 100% be able to stop any arbitrary container that they want to. As long as 'docker exec' works on managed containers, there is the possibility to need to 'docker stop' a particular managed container. Just like with a crashed container, you may want to stop and investigate, not just remove it.

**Comment by stevvooe** [comment]
Containers are managed by a task that must maintain an accurate state of the container. Tasks can only proceed in one direction, towards failure or success. As tasks proceed through their lifecycle, they report state that monotonically increases. If you startup containers that are managed as tasks, this guarantee is broken and the task will no longer manage container.

The lifecycle of containers managed by a service is set by the manager (via the dispatcher). The manager passes down the list of assigned tasks to a worker and the worker will hold on to those as long as they are there. This allows the manager to have some control over the lifecycle of the containers in the cluster to support investigation after a crash.

For the most, its okay to stop containers. They will be restarted by the orchestrator. If a container is restarted by another mechanism, the task controller won't try to shut it down again. This is a property of the system to handle cases like this. The issue here is that they won't be shutdown automatically, since the user has effectively taken control of them. If we didn't do this, we would shut it down immediately after you started it.

The current behavior is that these will go away when the service is deleted. For a long running service, this may be problematic but they can be pruned from the assignment set to keep it manageable.

__TL; DR__ If you take over the management of an individual container, swarm mode will no longer manage its lifecycle. It is up to the user to delete the extra container. In general, don't start containers managed by a service but there isn't a good reason to prevent it.

Also, there is a networking problem preventing the startup of the new containers, which I suspect is the issue with `3/5` replicas.

**Comment by mrjana** [comment]
@stevvooe I don't think we can allow containers to get decoupled from the manager lifecycle. Especially for networking reasons. Because they hold resources allocated by the manager. This is what manager-centric resource allocation results in. So either that container needs to get managed all the time even if they get manually restarted using an OOB mechanism or we should disallow those OOB mechanisms to interfere with swarm container lifecycle management.

**Comment by stevvooe** [comment]
@mrjana I agree that it should discouraged, but I am not sure if we should prevent it. Can we not cover this with documentation?

Ultimately, these containers will still be cleaned up when the assignments are removed. I've also been toying with the idea of a task reaper to really make sure these things are destroyed.

**Comment by thaJeztah** [comment]
@stevvooe @mrjana what should we do for 1.12? Should we add a filter to easily identify / filter managed/unmanaged containers and tell people "don't touch those managed ones", and open an issue for a "task reaper" in 1.13?

**Comment by tiborvass** [comment]
@thaJeztah i changed the title on this issue to make it clear.

**Comment by cirocosta** [comment]
Hey, are there any decisions regarding this? Or some discussion around it somewhere?

Thx!
